docs: Added documentation of Expression Monitors (#178)
scrapinghub/spidermon: docs/source/expression-monitors.rst

.. _topics-expression-monitors:

====================
Expression Monitors
====================

Expressions monitors are Monitors created on-the-fly when the Spidermon extension 
initializes.

Expression monitors can create tests based on simple expressions defined
in a dictionary in your settings like::

    SPIDERMON_SPIDER_OPEN_EXPRESSION_MONITORS = [
        {
            "name": "DumbChecksMonitor",
            "tests": [
                {
                    "name": "test_spider_name",
                    "expression": "spider.name == 'httpbin'",
                },
            ],
        },
    ]

The definition of each monitor should follow the :ref:`expression-monitor-schema`

> There are 3 settings related to the expression

Reviewer: Agreed. If we introduce more settings later, we may forgive to update this line.

monitors, they are:

    * :ref:`SPIDERMON_SPIDER_OPEN_EXPRESSION_MONITORS`
    * :ref:`SPIDERMON_SPIDER_CLOSE_EXPRESSION_MONITORS`
    * :ref:`SPIDERMON_ENGINE_STOP_EXPRESSION_MONITORS`

You have the following objects available to be used in your *expression*:
    
    * stats
    * crawler
    * spider
    * job
    * validation

>     * responses

Reviewer: We can leave for another PR, but if it not difficult, including some explanation about what can be accessed from these variables would be nice.

.. note::

    Notice that not the full-set of the Python Language features are available 
    to the expressions, only the ones that makes sense for a simple expressions
    that should evaluate to ``True`` or ``False``.

    To have a more deep understand about which features of the language are available

>     please refer to ``spidermon.python.intrepeter.Interpreter``.

Reviewer: Typo: spidermon.python.**interpreter**.Interpreter

Interpreter class needs testing and documentation as well, but it is not related to this PR :-)

>     please refer to ``spidermon.python.intrepeter.Interpreter``.

Author: I agree, but right now this will lead to a new topic at least.
I'll fix the typo for now.

.. _how-to-create-expression-monitor:

How to create a Expression Monitor
==================================

> First of all you should have :ref:`enabled spidermon <enabling-spidermon>`.

Reviewer: I agree. It can be removed.

Then you need to choose *when* you want to run your expression monitors

> You can use :ref:`SPIDERMON_SPIDER_OPEN_EXPRESSION_MONITORS` to run the monitor

Reviewer: When mentioning some settings constants, you could add a link to the section in settings.rst file.

> You can use :ref:`SPIDERMON_SPIDER_OPEN_EXPRESSION_MONITORS` to run the monitor

Author: It's linked @rennerocha

> when the spider open, or :ref:`SPIDERMON_SPIDER_CLOSE_EXPRESSION_MONITORS` if you

Reviewer 2: ```suggestion
when a spider opens, or :ref:`SPIDERMON_SPIDER_CLOSE_EXPRESSION_MONITORS` if you
```

> want to run when the spider is closed.

Reviewer 2: ```suggestion
want to run a monitor when a spider is closed.
```

> There's also the :ref:`SPIDERMON_ENGINE_STOP_EXPRESSION_MONITORS` to run the monitors

Reviewer 2: ```suggestion
There's also the :ref:`SPIDERMON_ENGINE_STOP_EXPRESSION_MONITORS` setting to run a monitor
```

The 1st monitor with 2 tests, one for checking the spider name and other test to
check if the crawler is there. The 2nd monitor will check if the spider 
finished with ``finished``::

    [
        {
            "name": "DumbChecksMonitor",
            "description": "My expression monitor",
            "tests": [
                {
                    "name": "test_spider_name",
                    "description": "Test spider name",
                    "expression": "spider.name == 'httpbin'",
                },
                {
                    "name": "test_crawler_exists",
                    "description": "Test Crawler exists",
                    "expression": "crawler is not None"
                }
            ],
        },
        {
            "name": "FinishedOkMonitor",
            "description": "My expression monitor 2",
            "tests": [
                {
                    "name": "test_finish_reason",
                    "description": "Test finish reason",
                    "expression": 'stats["finish_reason"] == "finished"',
                }
            ],
        }
    ]

.. _expression-monitor-schema:

> Expression Monitor Schema

Reviewer 2: :lipstick: 

```suggestion
Expression monitor schema
```

=========================

> Each `Expression Monitor` should follow this schema::

Reviewer 2: ```suggestion
Each `expression monitor` should follow this schema::
```
